wallet/txauthor/author.go

```diff
@@ -70,102 +73,140 @@ type ChangeSource struct {
 	ScriptSize int
 }
 
-// NewUnsignedTransaction creates an unsigned transaction paying to one or more
+// NewUnsignedTransaction creates an unsigned transaction paying to non or more
 // non-change outputs.  An appropriate transaction fee is included based on the
 // transaction size.
 //
-// Transaction inputs are chosen from repeated calls to fetchInputs with
-// increasing targets amounts.
+// Transaction inputs are added depending on the input selection strategy.
+// In general inputs are selected one at a time and evaluated whether
+// the current inputs suffice the funding amount plus fees.
+// When constant input selection is required all inputs are added in a batch
+// to be more efficient.
 //
-// If any remaining output value can be returned to the wallet via a change
-// output without violating mempool dust rules, a P2WPKH change output is
-// appended to the transaction outputs.  Since the change output may not be
-// necessary, fetchChange is called zero or one times to generate this script.
-// This function must return a P2WPKH script or smaller, otherwise fee estimation
-// will be incorrect.
+// When the input overshoots the amount of outputs an appropriate change is
+// constructed. This change output however is not considered when its below
+// the dust limit of the bitcoin network.
 //
 // If successful, the transaction, total input value spent, and all previous
-// output scripts are returned.  If the input source was unable to provide
-// enough input value to pay for every output any any necessary fees, an
+// output scripts are returned.  If the inputs were unable to provide
+// enough value to pay for every output and any necessary fees, an
 // InputSourceError is returned.
 //
-// BUGS: Fee estimation may be off when redeeming non-compressed P2PKH outputs.
+// BUGS: Fee estimation is off when redeeming inputs from a uncompressed P2PKH.
+// Because one cannot evaluate whether an uncrompressed or compressed public
+// key was used in the P2PKH before hashing the actual public effectively
+// knowing the detailed signature of such an input. Because umcompressed keys
+// are rare the worst case estimate for a P2PKH is a compressed public key.
 func NewUnsignedTransaction(outputs []*wire.TxOut, feeRatePerKb btcutil.Amount,
-	fetchInputs InputSource, changeSource *ChangeSource) (*AuthoredTx, error) {
+	inputs []wtxmgr.Credit, selectionStrategy InputSelectionStrategy,
```

Comment:
Question, can the signature change be a breaking change for the user?

Reply:
Not sure the `NewUnsignedTransaction` does not need any signatures for the Inputs, it estimates the `ScriptSig/RedeemScript` given worst case scenarios for differrent input types. So don't think it would be a problem

Reply:
Ah sorry, I was unclear, I meant the function signature :)

Reply:
Ahh yeah that's for sure, hmm I guess we need to introduce a `NewUnsignedTransaction` and slowly deprecate the old I guess ?